pkg/stub/handler.go

```diff
@@ -247,3 +255,179 @@ func (h *Handler) ensureRouterDeleted(ci *ingressv1alpha1.ClusterIngress) error
+
+		return
+	}
+
+	ns, ingresses, daemonsets, err := h.getOperatorState()
+	if err != nil {
+		logrus.Errorf("syncOperatorStatus: getOperatorState: %v", err)
+
+		return
+	}
+
+	oldCo := co
+	co = co.DeepCopy()
+	setStatusConditions(co, ns, ingresses, daemonsets)
+
+	if isNotFound {
+		co.Status.Version = operatorversion.Version
+
+		if err := sdk.Create(co); err != nil {
+			logrus.Errorf("syncOperatorStatus: failed to create ClusterOperator %s/%s: %v",
+				co.Namespace, co.Name, err)
+		} else {
+			logrus.Infof("syncOperatorStatus: created ClusterOperator %s/%s (UID %v)",
+				co.Namespace, co.Name, co.UID)
+		}
+
+		return
+	}
+
+	if clusteroperator.StatusEqual(&oldCo.Status, &co.Status) {
+		return
+	}
+
+	if err := sdk.Update(co); err != nil {
+		logrus.Errorf("syncOperatorStatus: failed to update status of ClusterOperator %s/%s: %v",
+			co.Namespace, co.Name, err)
+	}
+}
+
+// getOperatorState gets and returns the resources necessary to compute the
+// operator's current state.
+func (h *Handler) getOperatorState() (*corev1.Namespace, *ingressv1alpha1.ClusterIngressList, *appsv1.DaemonSetList, error) {
+	ns, err := h.manifestFactory.RouterNamespace()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"error building router namespace: %v", err)
+	}
+
+	if err := sdk.Get(ns); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil, nil, nil
+		}
+
+		return nil, nil, nil, fmt.Errorf(
+			"error getting Namespace %s: %v", ns.Name, err)
+	}
+
+	ingressList := &ingressv1alpha1.ClusterIngressList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterIngress",
+			APIVersion: "ingress.openshift.io/v1alpha1",
+		},
+	}
+	err = sdk.List(h.namespace, ingressList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list ClusterIngresses: %v", err)
+	}
+
+	daemonsetList := &appsv1.DaemonSetList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: "extensions/v1beta1",
+		},
+	}
+	err = sdk.List(h.namespace, daemonsetList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list DaemonSets: %v", err)
+	}
+
+	return ns, ingressList, daemonsetList, nil
+}
+
+// setStatusConditions sets the specified ClusterOperator's status using the
+// provided resources to compute the operator's current state.
+func setStatusConditions(co *osv1.ClusterOperator, ns *corev1.Namespace, ingressList *ingressv1alpha1.ClusterIngressList, daemonsetList *appsv1.DaemonSetList) {
+	failingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorFailing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ns == nil {
+		failingCondition.Status = osv1.ConditionTrue
+		failingCondition.Reason = "NoNamespace"
+		failingCondition.Message = "router namespace does not exist"
+	} else {
+		failingCondition.Status = osv1.ConditionFalse
+	}
+	clusteroperator.SetStatusCondition(co, failingCondition)
```

Comment:
Style issue: mutating input is troublesome. This function is close to stateless; how about just returning the new Status object based on immutable inputs?

Reply:
Fixed.